File: 36. Valid Sudoku.cpp
/*
Determine if a Sudoku is valid, according to: Sudoku Puzzles - The Rules.

The Sudoku board could be partially filled, where empty cells are filled with the character '.'.

A partially filled sudoku which is valid.
*/
class Solution {
public:
	bool isValidSudoku(vector<vector<char>>& board) {
		bool is_filled[9];
		int i, j;
		//check row
		for (i = 0; i < 9; i++) {
			fill(is_filled, is_filled + 9, false);
			for (j = 0; j < 9; j++) {
				if (board[i][j] == '.')
					continue;
				if (is_filled[board[i][j] - '1'])
					return false;

				if (!is_filled[board[i][j] - '1'])
					is_filled[board[i][j] - '1'] = true;
			}
		}

		//check column
		for (j = 0; j < 9; j++) {
			fill(is_filled, is_filled + 9, false);
			for (i = 0; i < 9; i++) {
				if (board[i][j] == '.')
					continue;
				if (is_filled[board[i][j] - '1'])
					return false;
				if (!is_filled[board[i][j] - '1'])
					is_filled[board[i][j] - '1'] = true;
			}
		}

		//check sub-boxes
		for (i = 0; i < 3; i++)
			for (j = 0; j < 3; j++) {
				fill(is_filled, is_filled + 9, false);
				for (int row = i * 3; row < i * 3 + 3; row++)
					for (int col = j * 3; col < j * 3 + 3; col++) {
						if (board[row][col] == '.')
							continue;
						if (is_filled[board[row][col] - '1'])
							return false;
						if (!is_filled[board[row][col] - '1'])
							is_filled[board[row][col] - '1'] = true;
					}
			}
		return true;
	}
};

File: Hash Table/36. Valid Sudoku.py
# Determine if a 9x9 Sudoku board is valid. Only the filled cells need to be validated according to the following rules:
#
# Each row must contain the digits 1-9 without repetition.
# Each column must contain the digits 1-9 without repetition.
# Each of the 9 3x3 sub-boxes of the grid must contain the digits 1-9 without repetition.
#
# A partially filled sudoku which is valid.
#
# The Sudoku board could be partially filled, where empty cells are filled with the character '.'.
#
# Example 1:
#
# Input:
# [
#   ["5","3",".",".","7",".",".",".","."],
#   ["6",".",".","1","9","5",".",".","."],
#   [".","9","8",".",".",".",".","6","."],
#   ["8",".",".",".","6",".",".",".","3"],
#   ["4",".",".","8",".","3",".",".","1"],
#   ["7",".",".",".","2",".",".",".","6"],
#   [".","6",".",".",".",".","2","8","."],
#   [".",".",".","4","1","9",".",".","5"],
#   [".",".",".",".","8",".",".","7","9"]
# ]
# Output: true
# Example 2:
#
# Input:
# [
#   ["8","3",".",".","7",".",".",".","."],
#   ["6",".",".","1","9","5",".",".","."],
#   [".","9","8",".",".",".",".","6","."],
#   ["8",".",".",".","6",".",".",".","3"],
#   ["4",".",".","8",".","3",".",".","1"],
#   ["7",".",".",".","2",".",".",".","6"],
#   [".","6",".",".",".",".","2","8","."],
#   [".",".",".","4","1","9",".",".","5"],
#   [".",".",".",".","8",".",".","7","9"]
# ]
# Output: false
# Explanation: Same as Example 1, except with the 5 in the top left corner being
#     modified to 8. Since there are two 8's in the top left 3x3 sub-box, it is invalid.
# Note:
#
# A Sudoku board (partially filled) could be valid but is not necessarily solvable.
# Only the filled cells need to be validated according to the mentioned rules.
# The given board contain only digits 1-9 and the character '.'.
# The given board size is always 9x9.

class Solution(object):
    def isValidSudoku(self, board):
        """
        :type board: List[List[str]]
        :rtype: bool
        """
        rows = [set() for _ in range(9)]
        cols = [set() for _ in range(9)]
        squares = [set() for _ in range(9)]

        for i in range(9):
            for j in range(9):
                if board[i][j] != '.':
                    numSquare = i // 3 * 3 + j // 3
                    if board[i][j] in rows[i] or board[i][j] in cols[j] or board[i][j] in squares[numSquare]:
                        return False
                    rows[i].add(board[i][j])
                    cols[j].add(board[i][j])
                    squares[numSquare].add(board[i][j])

        return True
